packages/runner/src/run.ts

```diff
@@ -234,20 +235,31 @@ export async function runTest(test: Test | Custom, runner: VitestRunner) {
   updateTask(test, runner)
 }
 
-function failTask(result: TaskResult, err: unknown, diffOptions?: DiffOptions) {
+function failTask(result: TaskResult, err: unknown, diffOptions?: DiffOptions): ErrorWithDiff[] {
   if (err instanceof PendingError) {
     result.state = 'skip'
-    return
+    return []
   }
 
   result.state = 'fail'
   const errors = Array.isArray(err)
     ? err
     : [err]
-  for (const e of errors) {
-    const error = processError(e, diffOptions)
-    result.errors ??= []
-    result.errors.push(error)
+  const processedErrors = errors.map(e => processError(e, diffOptions))
+  result.errors = [...result.errors ?? [], ...processedErrors]
+  return processedErrors
+}
+
+function markTasksAsFailed(suite: Suite, errors: ErrorWithDiff[], runner: VitestRunner) {
+  for (const t of getTests(suite)) {
```

Comment:
I don't like doing this for every test because the terminal seems bloated. We have a `suite.result` and it's enough. I think reporters should handle this case if they don't distinguish between suite fails and test fails

<img width="459" alt="Screenshot 2023-12-28 at 11 20 33" src="https://github.com/vitest-dev/vitest/assets/16173870/042c4210-8243-40b3-9886-da56a7ffa9d6">

Reply:
> I don't like doing this for every test because the terminal seems bloated

That sounds fair. To explain my perspective, the reason I went with this is that I thought `beforeAll/afterAll` failures should be more aligned with `beforeEach/afterEach` failures.

I experimented with current behavior on stackblitz:

https://stackblitz.com/edit/vitest-dev-vitest-m7bw4w?file=test%2Frepro.test.ts

<details><summary>Vitest UI screenshot</summary>

![image](https://github.com/vitest-dev/vitest/assets/4232207/0ba7cb49-5d5e-4810-a316-0b575f101406)

</details>

For example, currently `afterAll` failure would keep tests as success and I thought that's not an intended behavior.

To avoid cluttering terminal with a single error appearing in all tests, I think I could do something like this while still flagging tests as "fail":

```tsx
      t.result = {
        ...t.result,
        // flag as "fail"
        state: 'fail',
        // but don't have to copy errors
        // errors: [...t.result?.errors ?? [], ...errors],
      }
```

What do you think?

Reply:
>  To explain my perspective, the reason I went with this is that I thought `beforeAll`/`afterAll` failures should be more aligned with `beforeEach`/`afterEach` failures.

`afterAll`/`beforeAll` are not bound to any test, they are suite hooks. `beforeEach`/`afterEach` are test hooks. What I would expect is for tests to have `skip` state if `beforeAll` failed (because we don't run any test), and any state if `afterAll` failed (only the suite is marked as failed)

> For example, currently afterAll failure would keep tests as success and I thought that's not an intended behavior.

`afterAll` fails only a suite which seems correct to me.

Reply:
> afterAll fails only a suite which seems correct to me.

This is consistent with `mocha` for example:

```ts
import { test, after } from 'mocha'

after(() => {
  throw new Error('after')
})

test('test', () => {})
```

Although in mocha `afterEach` doesn't fail the test while in Vitest it does.

If we are making it consistent, then 1) this is definitely a breaking change, 2) we need to decide what is consistent

Reply:
> afterAll/beforeAll are not bound to any test, they are suite hooks. beforeEach/afterEach are test hooks.

Totally good point. I didn't have that understanding. 

Okay, I'll see what I can do with junit reporter. Probably I'll close this PR and create a new one.

Thanks again for the review!